### ISSUE TITLE
Ambassador pods crash when frequently creating/deleting mappings

### ISSUE THREAD

**Issue by juanjoku** [datawire/ambassador#3280]
**Describe the bug**
As part of our testing, one of the tests does the following:

- There is one single TLSContext, using a wildcard certificate.
- Infinite loop
  - 50 mappings are created (1sec delay between each operation)
  - Then, the 50 mappings are deleted (1sec delay between each operation)
- During the test, there is no traffic for these mappings. There are other mappings receiving traffic (few requests, minimal load).

With Ambasador 1.12.1, after some minutes (3 to 10 in our tests), Ambassador pods die. This test passed with Ambassador 1.11.1.

**Additional context**
When the pod fails, Ambassador (loglevel=trace) has logged this error:

```
[...]
time="2021-03-18 14:57:12" level=debug msg="Taking K8sSecret snapshot.SecretRef{Namespace:\"kumori\", Name:\"cluster-core-cert-server-wildcard-test-kumori-cloud-cert\"}" func=github.com/datawire/ambassador/cmd/entrypoint.ReconcileSecrets file="github.com/datawire/ambassador/cmd/entrypoint/secrets.go:140" CMD=entrypoint PID=1 THREAD=/watcher
time="2021-03-18 14:57:12" level=debug msg="Taking K8sSecret snapshot.SecretRef{Namespace:\"kumori\", Name:\"kumori-examples-cert-server-wildcard-vera-kumori-cloud-cert\"}" func=github.com/datawire/ambassador/cmd/entrypoint.ReconcileSecrets file="github.com/datawire/ambassador/cmd/entrypoint/secrets.go:140" CMD=entrypoint PID=1 THREAD=/watcher
time="2021-03-18 14:57:12" level=debug msg="Taking K8sSecret snapshot.SecretRef{Namespace:\"kumori\", Name:\"wildcard-default\"}" func=github.com/datawire/ambassador/cmd/entrypoint.ReconcileSecrets file="github.com/datawire/ambassador/cmd/entrypoint/secrets.go:140" CMD=entrypoint PID=1 THREAD=/watcher
time="2021-03-18 14:57:12" level=debug msg="WATCHER: KubernetesEndpointResolver resolver endpoint-alertmanager is active (CRD)" func="github.com/datawire/ambassador/cmd/entrypoint.(*endpointRoutingInfo).saveResolver" file="github.com/datawire/ambassador/cmd/entrypoint/endpoints.go:201" CMD=entrypoint PID=1 THREAD=/watcher
time="2021-03-18 14:57:12" level=debug msg="WATCHER: KubernetesEndpointResolver resolver endpoint-grafana is active (CRD)" func="github.com/datawire/ambassador/cmd/entrypoint.(*endpointRoutingInfo).saveResolver" file="github.com/datawire/ambassador/cmd/entrypoint/endpoints.go:201" CMD=entrypoint PID=1 THREAD=/watcher
time="2021-03-18 14:57:12" level=debug msg="WATCHER: KubernetesEndpointResolver resolver endpoint-prometheus is active (CRD)" func="github.com/datawire/ambassador/cmd/entrypoint.(*endpointRoutingInfo).saveResolver" file="github.com/datawire/ambassador/cmd/entrypoint/endpoints.go:201" CMD=entrypoint PID=1 THREAD=/watcher
time="2021-03-18 14:57:12" level=debug msg="WATCHER: KubernetesEndpointResolver resolver kumori-default is active (CRD)" func="github.com/datawire/ambassador/cmd/entrypoint.(*endpointRoutingInfo).saveResolver" file="github.com/datawire/ambassador/cmd/entrypoint/endpoints.go:201" CMD=entrypoint PID=1 THREAD=/watcher
time="2021-03-18 14:57:12" level=debug msg="WATCHER: endpoint resolver endpoint exists by default" func="github.com/datawire/ambassador/cmd/entrypoint.(*endpointRoutingInfo).reconcileEndpointWatches" file="github.com/datawire/ambassador/cmd/entrypoint/endpoints.go:116" CMD=entrypoint PID=1 THREAD=/watcher
time="2021-03-18 14:57:12" level=debug msg="WATCHER: endpoint resolver kubernetes-endpoint exists by default" func="github.com/datawire/ambassador/cmd/entrypoint.(*endpointRoutingInfo).reconcileEndpointWatches" file="github.com/datawire/ambassador/cmd/entrypoint/endpoints.go:116" CMD=entrypoint PID=1 THREAD=/watcher
time="2021-03-18 14:57:12" level=debug msg="WATCHER: TCPMapping kt-082051-eedebd49-19031-tcpmapping-linked uses the default resolver (CRD)" func="github.com/datawire/ambassador/cmd/entrypoint.(*endpointRoutingInfo).checkTCPMapping" file="github.com/datawire/ambassador/cmd/entrypoint/endpoints.go:232" CMD=entrypoint PID=1 THREAD=/watcher
time="2021-03-18 14:57:12" level=debug msg="WATCHER: --------" func=github.com/datawire/ambassador/cmd/entrypoint.watcherLoop file="github.com/datawire/ambassador/cmd/entrypoint/watcher.go:192" CMD=entrypoint PID=1 THREAD=/watcher
time="2021-03-18 14:57:12" level=debug msg="WATCHER: --------" func=github.com/datawire/ambassador/cmd/entrypoint.watcherLoop file="github.com/datawire/ambassador/cmd/entrypoint/watcher.go:192" CMD=entrypoint PID=1 THREAD=/watcher
panic: reflect: slice index out of range [recovered]
	panic: reflect: slice index out of range

goroutine 1798 [running]:
encoding/json.(*encodeState).marshal.func1(0xc00108bd88)
	encoding/json/encode.go:326 +0x85
panic(0x1b16920, 0x20bc530)
	runtime/panic.go:969 +0x175
reflect.Value.Index(0x1ab7440, 0xc00073f938, 0x197, 0x7, 0x1e50100, 0xc0018bcf30, 0x196)
	reflect/value.go:961 +0x1f2
encoding/json.arrayEncoder.encode(0xc000e5dd70, 0xc000f5c080, 0x1ab7440, 0xc00073f938, 0x197, 0x100)
	encoding/json/encode.go:886 +0x8d
encoding/json.sliceEncoder.encode(0xc000e5dd90, 0xc000f5c080, 0x1ab7440, 0xc00073f938, 0x197, 0x100)
	encoding/json/encode.go:860 +0x8f
encoding/json.structEncoder.encode(0xc0002f8600, 0x15, 0x21, 0xc000ace660, 0xc000f5c080, 0x1e4d9e0, 0xc00073f8c0, 0x199, 0xc000f00100)
	encoding/json/encode.go:759 +0x2ab
encoding/json.ptrEncoder.encode(0xc000ace690, 0xc000f5c080, 0x1bbd420, 0xc001a41388, 0x196, 0x100)
	encoding/json/encode.go:915 +0x125
encoding/json.structEncoder.encode(0xc00079c480, 0x5, 0x8, 0xc000ace9c0, 0xc000f5c080, 0x1d441c0, 0xc001a41380, 0x199, 0x1b90100)
	encoding/json/encode.go:759 +0x2ab
encoding/json.ptrEncoder.encode(0xc000ace9f0, 0xc000f5c080, 0x1b7e480, 0xc001a41380, 0x16, 0x1b70100)
	encoding/json/encode.go:915 +0x125
encoding/json.(*encodeState).reflectValue(0xc000f5c080, 0x1b7e480, 0xc001a41380, 0x16, 0xc001080100)
	encoding/json/encode.go:358 +0x82
encoding/json.(*encodeState).marshal(0xc000f5c080, 0x1b7e480, 0xc001a41380, 0x7f0d6b490100, 0x0, 0x0)
	encoding/json/encode.go:330 +0xf4
encoding/json.Marshal(0x1b7e480, 0xc001a41380, 0xc0008ce510, 0xc001a41301, 0x0, 0x0, 0xc00108be58)
	encoding/json/encode.go:161 +0x52
encoding/json.MarshalIndent(0x1b7e480, 0xc001a41380, 0x0, 0x0, 0x1e898fa, 0x2, 0xc000b80ed0, 0xc00118ef00, 0xc0012f4f88, 0xc00108bf60, ...)
	encoding/json/encode.go:176 +0x3f
github.com/datawire/ambassador/cmd/entrypoint.(*SnapshotHolder).Notify(0xc0008ce510, 0x21346c0, 0xc000b80ed0, 0xc000729540, 0xc00118f6e0, 0xc000ba59d0)
	github.com/datawire/ambassador/cmd/entrypoint/watcher.go:430 +0x1ad
github.com/datawire/ambassador/cmd/entrypoint.watcherLoop.func1(0xc000c25980, 0x21346c0, 0xc000b80ed0, 0xc000729540, 0xc00118f6e0, 0xc000ba59d0)
	github.com/datawire/ambassador/cmd/entrypoint/watcher.go:184 +0x85
created by github.com/datawire/ambassador/cmd/entrypoint.watcherLoop
	github.com/datawire/ambassador/cmd/entrypoint/watcher.go:180 +0x1f1
```

Used mappings:
```
apiVersion: getambassador.io/v2
kind: Mapping
metadata:
  name: mymapping-i // i = [1,2,....,100]
  namespace: kumori
spec:
  add_request_headers: {}
  host: mytest-i.mycompany.com // i = [1,2,....,100]
  load_balancer:
    policy: round_robin
  precedence: 1
  prefix: /.*.*
  prefix_regex: true
  resolver: kumori-default
  rewrite: ""
  service: myservice.kumori:5000 // The same service for all the mappings
  timeout_ms: 3600000
  use_websocket: false
```

Ambassador Daemonset:
```
apiVersion: apps/v1
kind: DaemonSet
metadata:
  annotations:
    deprecated.daemonset.template.generation: "7"
  name: ambassador
  namespace: kumori
spec:
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      service: ambassador
  template:
    metadata:
      annotations:
        consul.hashicorp.com/connect-inject: "false"
        kubectl.kubernetes.io/restartedAt: "2021-03-18T12:44:12+01:00"
        sidecar.istio.io/inject: "false"
      creationTimestamp: null
      labels:
        app.kubernetes.io/managed-by: getambassador.io
        service: ambassador
        type: ingress
    spec:
      containers:
      - env:
        - name: AMBASSADOR_SINGLE_NAMESPACE
        - name: AMBASSADOR_DRAIN_TIME
          value: "600"
        - name: AMBASSADOR_FAST_RECONFIGURE
          value: "true"
        - name: AES_LOG_LEVEL
          value: info
        - name: AMBASSADOR_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        - name: HOST_IP
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: status.hostIP
        image: docker.io/datawire/ambassador:1.12.1
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /ambassador/v0/check_alive
            port: 8877
            scheme: HTTP
          initialDelaySeconds: 30
          periodSeconds: 3
          successThreshold: 1
          timeoutSeconds: 1
        name: ambassador
        ports:
        - containerPort: 8443
          hostPort: 8443
          name: https
          protocol: TCP
        - containerPort: 8877
          hostPort: 8080
          name: admin
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /ambassador/v0/check_ready
            port: 8877
            scheme: HTTP
          initialDelaySeconds: 30
          periodSeconds: 3
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            cpu: "1"
            memory: 800Mi
          requests:
            cpu: 200m
            memory: 300Mi
        securityContext:
          allowPrivilegeEscalation: false
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /tmp/ambassador-pod-info
          name: ambassador-pod-info
      dnsPolicy: ClusterFirstWithHostNet
      nodeSelector:
        kumori/ingress: "true"
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext:
        runAsUser: 8888
      serviceAccount: ambassador
      serviceAccountName: ambassador
      terminationGracePeriodSeconds: 30
      volumes:
      - downwardAPI:
          defaultMode: 420
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.labels
            path: labels
        name: ambassador-pod-info
  updateStrategy:
    rollingUpdate:
      maxUnavailable: 1
    type: RollingUpdate
```

**Comment by juanjoku** [comment]
Has no one else detected this problem?

**Comment by juanjoku** [comment]
Hello,
I have repeated the test using docker.io/datawiredev/ambassador:8f0a64551e11, and it works fine  :-)

**Comment by juanjoku** [comment]
@rhs, do you know if this correction will be included in the next release?   
If so, will it be in the short term?

Thank you very much!

**Comment by juanjoku** [comment]
Hello @rhs, 

I have repeated the test using Ambassador v1.13.1, and it is working fine.
So... v1.13.1 includes the correction?

Thx.

**Comment by juanjoku** [comment]
We were informed via Slack (by Cindy Mullins) that the fix was included en v1.13.0. We tried to reproduce the issue using 1.13 and the issue seems to be solved.
